File: apps/gig-crawler-2/src/adapters/GeminiRepo/GeminiAdapter.ts
import { GoogleGenerativeAI } from "@google/generative-ai";
import type { LLMPort } from "../../ports/LLMPort.js";
import type { SearchResult } from "../../models/searchResult.js";
import type { ScrapedContent } from "../../models/scrapedContent.js";
import type { Gig } from "../../models/gig.js";
import { logger } from "../../utils/logger.js";
import { retry } from "../../utils/retry.js";
import { parseFlexibleDate } from "../../utils/dateUtils.js";
import { URL_FILTER_PROMPT } from "../../prompts/urlFilter.js";
import { GIG_EXTRACTION_PROMPT } from "../../prompts/gigExtraction.js";
import { GIG_EXTRACTION_BATCH_PROMPT } from "../../prompts/gigExtractionBatch.js";
import { env } from "../../models/env.js";

export class GeminiAdapter implements LLMPort {
  private readonly genAI: GoogleGenerativeAI;
  private readonly model: string;

  constructor(apiKey: string = env.GEMINI_API_KEY, model: string = env.GEMINI_MODEL) {
    this.genAI = new GoogleGenerativeAI(apiKey);
    this.model = model;
  }

  async filterPromisingUrls(
    searchResults: SearchResult[]
  ): Promise<string[]> {
    return retry(
      async () => {
        logger.info(
          { count: searchResults.length },
          "Filtering search results with Gemini"
        );

        const resultsText = searchResults
          .map(
            (r) =>
              `URL: ${r.url}\nTitle: ${r.title}\nDescription: ${r.description || "N/A"}`
          )
          .join("\n\n");

        const model = this.genAI.getGenerativeModel({
          model: this.model,
          generationConfig: {
            temperature: 0.1,
            responseMimeType: "application/json",
          },
        });

        const prompt = URL_FILTER_PROMPT(resultsText);
        const result = await model.generateContent(prompt);
        const response = result.response;
        const text = response.text();

        try {
          const data = JSON.parse(text);
          const urls = data.promising_urls || [];
          logger.info({ count: urls.length }, "Filtered to promising URLs");
          return urls;
        } catch (error) {
          logger.error({ error, text }, "Failed to parse Gemini response");
          return [];
        }
      },
      {
        maxAttempts: 3,
        onError: (error, attempt) => {
          logger.warn({ error, attempt }, "Gemini filter attempt failed");
        },
      }
    );
  }

  async extractGigsFromContent(
    scrapedContent: ScrapedContent
  ): Promise<Gig[]> {
    if (!scrapedContent.success || !scrapedContent.text) {
      logger.warn(
        { url: scrapedContent.url },
        "Skipping extraction (no text)"
      );
      return [];
    }

    return retry(
      async () => {
        logger.info({ url: scrapedContent.url }, "Extracting gigs with Gemini");

        // Truncate text if too long
        const text = scrapedContent.text!.slice(0, 50000);

        const model = this.genAI.getGenerativeModel({
          model: this.model,
          generationConfig: {
            temperature: 0.1,
            responseMimeType: "application/json",
          },
        });

        const prompt = GIG_EXTRACTION_PROMPT(scrapedContent.url, text);
        const result = await model.generateContent(prompt);
        const response = result.response;
        const responseText = response.text();

        try {
          const data = JSON.parse(responseText);
          const gigsData = data.gigs || [];

          const gigs: Gig[] = [];
          for (const gigData of gigsData) {
            try {
              const date = parseFlexibleDate(gigData.date);
              if (!date) {
                logger.warn(
                  { gig: gigData },
                  "Skipping gig with invalid date"
                );
                continue;
              }

              gigs.push({
                title: gigData.title,
                date,
                venueName: gigData.venue_name || "Unknown Venue",
                description: gigData.description,
                price: gigData.price,
                url: scrapedContent.url,
                imageUrl: gigData.image_url,
              });
            } catch (error) {
              logger.warn({ error, gigData }, "Failed to parse gig");
            }
          }

          logger.info(
            { count: gigs.length, url: scrapedContent.url },
            "Extracted gigs"
          );
          return gigs;
        } catch (error) {
          logger.error(
            { error, text: responseText },
            "Failed to parse Gemini response"
          );
          return [];
        }
      },
      {
        maxAttempts: 3,
        onError: (error, attempt) => {
          logger.warn({ error, attempt }, "Gemini extraction attempt failed");
        },
      }
    );
  }

  async extractGigsFromMultiplePages(
    scrapedContents: ScrapedContent[]
  ): Promise<Gig[]> {
    const successfulPages = scrapedContents.filter(
      (sc) => sc.success && sc.text
    );

    if (successfulPages.length === 0) {
      logger.warn("No successful pages to extract from");
      return [];
    }

    return retry(
      async () => {
        logger.info(
          { pageCount: successfulPages.length },
          "Extracting gigs from multiple pages with Gemini (batch)"
        );

        // Prepare pages data
        const pages = successfulPages.map((sc) => ({
          url: sc.url,
          content: sc.text!.slice(0, 15000), // Limit per page
        }));

        const model = this.genAI.getGenerativeModel({
          model: this.model,
          generationConfig: {
            temperature: 0.1,
            responseMimeType: "application/json",
          },
        });

        const prompt = GIG_EXTRACTION_BATCH_PROMPT(pages);
        const result = await model.generateContent(prompt);
        const response = result.response;
        const responseText = response.text();

        try {
          const data = JSON.parse(responseText);
          const gigsData = data.gigs || [];

          const gigs: Gig[] = [];
          for (const gigData of gigsData) {
            try {
              const date = parseFlexibleDate(gigData.date);
              if (!date) {
                logger.warn(
                  { gig: gigData },
                  "Skipping gig with invalid date"
                );
                continue;
              }

              gigs.push({
                title: gigData.title,
                date,
                venueName: gigData.venue_name || "Unknown Venue",
                description: gigData.description,
                price: gigData.price,
                url: gigData.url || "", // LLM should provide source page URL
                imageUrl: gigData.image_url,
              });
            } catch (error) {
              logger.warn({ error, gigData }, "Failed to parse gig");
            }
          }

          logger.info(
            { count: gigs.length, pageCount: successfulPages.length },
            "Extracted gigs from all pages"
          );
          return gigs;
        } catch (error) {
          logger.error(
            { error, text: responseText },
            "Failed to parse Gemini batch response"
          );
          return [];
        }
      },
      {
        maxAttempts: 3,
        onError: (error, attempt) => {
          logger.warn({ error, attempt }, "Gemini batch extraction attempt failed");
        },
      }
    );
  }
}

File: apps/gig-crawler-2-python-backup/src/adapters/llm/gemini_adapter.py
"""Google Gemini LLM adapter."""
import json
import google.generativeai as genai
from typing import List
from ...ports.llm_port import LLMPort
from ...models.search_result import SearchResult
from ...models.scraped_content import ScrapedContent
from ...models.gig import Gig
from ...utils.logger import setup_logger
from ...utils.retry import async_retry

logger = setup_logger(__name__)


class GeminiLLMAdapter(LLMPort):
    """Google Gemini LLM implementation."""

    def __init__(self, api_key: str, model: str = "gemini-1.5-flash"):
        """Initialize Gemini client.

        Args:
            api_key: Google Gemini API key
            model: Model name (gemini-1.5-flash or gemini-1.5-pro)
        """
        genai.configure(api_key=api_key)
        self.model = genai.GenerativeModel(
            model,
            generation_config={
                "temperature": 0.1,
                "response_mime_type": "application/json",
            },
        )

    @async_retry(max_attempts=3, exceptions=(Exception,))
    async def filter_promising_urls(
        self, search_results: List[SearchResult]
    ) -> List[str]:
        """Filter search results to promising URLs (Pass 1).

        Args:
            search_results: List of search results from web search

        Returns:
            List of promising URLs to scrape
        """
        from ...prompts.url_filter import URL_FILTER_PROMPT

        logger.info(f"Filtering {len(search_results)} search results with Gemini")

        # Format search results for prompt
        results_text = "\n\n".join(
            f"URL: {r.url}\nTitle: {r.title}\nDescription: {r.description or 'N/A'}"
            for r in search_results
        )

        prompt = URL_FILTER_PROMPT.format(search_results=results_text)

        response = await self.model.generate_content_async(prompt)

        try:
            data = json.loads(response.text)
            urls = data.get("promising_urls", [])
            logger.info(f"Filtered to {len(urls)} promising URLs")
            return urls
        except json.JSONDecodeError as e:
            logger.error(f"Failed to parse Gemini response: {e}")
            logger.error(f"Response text: {response.text}")
            return []

    @async_retry(max_attempts=3, exceptions=(Exception,))
    async def extract_gigs_from_content(
        self, scraped_content: ScrapedContent
    ) -> List[Gig]:
        """Extract structured gig data from scraped content (Pass 2).

        Args:
            scraped_content: Scraped content from a URL

        Returns:
            List of extracted gigs
        """
        from ...prompts.gig_extraction import GIG_EXTRACTION_PROMPT

        if not scraped_content.success or not scraped_content.text:
            logger.warning(f"Skipping extraction for {scraped_content.url} (no text)")
            return []

        logger.info(f"Extracting gigs from {scraped_content.url}")

        # Truncate text if too long (to avoid token limits)
        text = scraped_content.text[:50000]

        prompt = GIG_EXTRACTION_PROMPT.format(
            url=scraped_content.url,
            content=text,
        )

        response = await self.model.generate_content_async(prompt)

        try:
            data = json.loads(response.text)
            gigs_data = data.get("gigs", [])

            gigs = []
            for gig_data in gigs_data:
                try:
                    # Parse date
                    from ...utils.date_utils import parse_flexible_date
                    date = parse_flexible_date(gig_data.get("date", ""))

                    if not date:
                        logger.warning(f"Skipping gig with invalid date: {gig_data}")
                        continue

                    gig = Gig(
                        title=gig_data.get("title", ""),
                        date=date,
                        venue_name=gig_data.get("venue_name", "Unknown Venue"),
                        description=gig_data.get("description"),
                        price=gig_data.get("price"),
                        url=scraped_content.url,
                        image_url=gig_data.get("image_url"),
                    )
                    gigs.append(gig)
                except Exception as e:
                    logger.warning(f"Failed to parse gig: {e} - {gig_data}")
                    continue

            logger.info(f"Extracted {len(gigs)} gigs from {scraped_content.url}")
            return gigs

        except json.JSONDecodeError as e:
            logger.error(f"Failed to parse Gemini response: {e}")
            logger.error(f"Response text: {response.text}")
            return []
